### PR TITLE
Adding more information on key location to error messages

### PR DESCRIPTION
**Pull request by mafurman**
NOTE: This is set to be be merged into the mafurman/lkg_refresh_on_network_error branch.

In the case where we fail to validate a signature on a token, this PR provides additional information for debugging purposes. If a 'kid' matched, the exception thrown will contain information on where this key was located (Config or TokenValidationParameters) as well as the number of keys on the Config and the TokenValidationParameters.

If the 'kid' did not match, only the number of keys on the Config and the TokenValidationParameters will be included in the resulting exception.

Currently, information on where a key came from in the successful validation case is not logged. This is done for performance reasons, but can be added in if necessary.

### CODE REVIEW

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs [hunk 1]
```diff
@@ -1321,10 +1321,21 @@ private static JsonWebToken ValidateSignature(string token, TokenValidationParam
...
+            var keysInConfiguration = TokenUtilities.GetAllSigningKeys(configuration);
+            var numKeysInTokenValidationParameters = keysInTokenValidationParameters.Count();
+            var numKeysInConfiguration = keysInConfiguration.Count();
+
             if (kidExists)
             {
                 if (kidMatched)
-                    throw LogHelper.LogExceptionMessage(new SecurityTokenInvalidSignatureException(LogHelper.FormatInvariant(TokenLogMessages.IDX10511, keysAttempted, jwtToken.Kid, exceptionStrings, jwtToken)));
+                {
+                    var isKidInTVP = keysInTokenValidationParameters.Any(x => x.KeyId.Equals(jwtToken.Kid, StringComparison.Ordinal));
```
Comment (lechu445) [inline src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs:1296]: why not simplify
```suggestion
                    var isKidInTVP = keysInTokenValidationParameters.Any(x => x.KeyId == jwtToken.Kid);
```

File: src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs [hunk 2]
```diff
@@ -1022,10 +1023,21 @@ protected virtual JwtSecurityToken ValidateSignature(string token, TokenValidati
...
+            // in the Configuration as well.
+            var keysInTokenValidationParameters = TokenUtilities.GetAllSigningKeys(validationParameters);
+            var numKeysInTokenValidationParameters = keysInTokenValidationParameters.Count();
+
             if (kidExists)
             {
                 if (kidMatched)
-                    throw LogHelper.LogExceptionMessage(new SecurityTokenInvalidSignatureException(LogHelper.FormatInvariant(TokenLogMessages.IDX10511, keysAttempted, jwtToken.Header.Kid, exceptionStrings, jwtToken)));
+                {
+                    var isKidInTVP = keysInTokenValidationParameters.Any(x => x.KeyId.Equals(jwtToken.Header.Kid, StringComparison.Ordinal));
```
Comment (lechu445) [inline src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs:1036]:
```suggestion
                    var isKidInTVP = keysInTokenValidationParameters.Any(x => x.KeyId == jwtToken.Header.Kid);
```

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs (hunk 1 above)
Comment (mafurman) [reply src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs:1296]: This allows us to specify the type of StringComparison being used, and so it's more explicit about how the strings are being compared. 

Also, I believe we generally use 'Equals' over '==' in this library even when the types of the variables being compared are strings.

**Comment by mafurman** [comment]
@RojaEnnam I got rid of duplicate code in AadIssuerValidator here: https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1786/commits/af2bfb19451a8b5df67a8f0be096979a8732429a

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs (hunk 1 above)
Comment (lechu445) [reply src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs:1296]: @mafurman thank you for the explanation.

I have proposed a separate PR https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1775  to change the convention as operator == makes more sense for me

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs [hunk 3]
```diff
@@ -1321,10 +1321,21 @@ private static JsonWebToken ValidateSignature(string token, TokenValidationParam
...
+                    var keyLocation = isKidInTVP ? "TokenValidationParameters" : "Configuration";
+                    throw LogHelper.LogExceptionMessage(new SecurityTokenInvalidSignatureException(
+                        LogHelper.FormatInvariant(TokenLogMessages.IDX10511, keysAttempted, LogHelper.MarkAsNonPII(numKeysInTokenValidationParameters), LogHelper.MarkAsNonPII(numKeysInConfiguration), LogHelper.MarkAsNonPII(keyLocation), jwtToken.Kid, exceptionStrings, jwtToken)));
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs]: for code with a large number of parameters, it reads easier if they are stacked, for example
```cs
 LogHelper.FormatInvariant(TokenLogMessages.IDX10511,
    keysAttempted, 
    LogHelper.MarkAsNonPII(numKeysInTokenValidationParameters), 
    LogHelper.MarkAsNonPII(numKeysInConfiguration),
    LogHelper.MarkAsNonPII(keyLocation), jwtToken.Kid, 
    exceptionStrings, 
    jwtToken)));

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs [hunk 4]
```diff
@@ -1336,11 +1347,13 @@ private static JsonWebToken ValidateSignature(string token, TokenValidationParam
...
                     configuration,
-                    exceptionStrings);
+                    exceptionStrings,
+                    numKeysInTokenValidationParameters,
+                    numKeysInConfiguration);
             }
 
             if (keysAttempted.Length > 0)
-                throw LogHelper.LogExceptionMessage(new SecurityTokenSignatureKeyNotFoundException(LogHelper.FormatInvariant(TokenLogMessages.IDX10503, keysAttempted, exceptionStrings, jwtToken)));
+                throw LogHelper.LogExceptionMessage(new SecurityTokenSignatureKeyNotFoundException(LogHelper.FormatInvariant(TokenLogMessages.IDX10503, keysAttempted, LogHelper.MarkAsNonPII(numKeysInTokenValidationParameters), LogHelper.MarkAsNonPII(numKeysInConfiguration), exceptionStrings, jwtToken)));
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs]: consider stacking parameters as above.

File: src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs [hunk 5]
```diff
@@ -1022,10 +1023,21 @@ protected virtual JwtSecurityToken ValidateSignature(string token, TokenValidati
...
+                    var keyLocation = isKidInTVP ? "TokenValidationParameters" : "Configuration";
+                    throw LogHelper.LogExceptionMessage(new SecurityTokenInvalidSignatureException(
+                        LogHelper.FormatInvariant(TokenLogMessages.IDX10511, keysAttempted, LogHelper.MarkAsNonPII(numKeysInTokenValidationParameters), LogHelper.MarkAsNonPII(0), LogHelper.MarkAsNonPII(keyLocation), jwtToken.Header.Kid, exceptionStrings, jwtToken)));
```
Comment (brentschmaltz) [inline src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs]: consider stacking parameters as above.

File: src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs [hunk 6]
```diff
@@ -1037,11 +1049,13 @@ protected virtual JwtSecurityToken ValidateSignature(string token, TokenValidati
...
                     null,
-                    exceptionStrings);
+                    exceptionStrings,
+                    numKeysInTokenValidationParameters,
+                    0);
             }
 
             if (keysAttempted.Length > 0)
-                throw LogHelper.LogExceptionMessage(new SecurityTokenSignatureKeyNotFoundException(LogHelper.FormatInvariant(TokenLogMessages.IDX10503, keysAttempted, exceptionStrings, jwtToken)));
+                throw LogHelper.LogExceptionMessage(new SecurityTokenSignatureKeyNotFoundException(LogHelper.FormatInvariant(TokenLogMessages.IDX10503, keysAttempted, LogHelper.MarkAsNonPII(numKeysInTokenValidationParameters), LogHelper.MarkAsNonPII(0), exceptionStrings, jwtToken)));
```
Comment (brentschmaltz) [inline src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs]: consider stacking parameters as above.

File: src/Microsoft.IdentityModel.Tokens/InternalValidators.cs [hunk 7]
```diff
@@ -85,7 +85,7 @@ internal static void ValidateLifetimeAndIssuerAfterSignatureNotValidatedJwt(Secu
 
                 throw LogHelper.LogExceptionMessage(new SecurityTokenUnableToValidateException(
                     validationFailure,
-                    LogHelper.FormatInvariant(TokenLogMessages.IDX10516, kid, exceptionStrings, securityToken, LogHelper.MarkAsNonPII(validLifetime), validIssuer)));
+                    LogHelper.FormatInvariant(TokenLogMessages.IDX10516, kid, LogHelper.MarkAsNonPII(numKeysInTokenValidationParameters), LogHelper.MarkAsNonPII(numKeysInConfiguration), exceptionStrings, securityToken, LogHelper.MarkAsNonPII(validLifetime), validIssuer)));
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/InternalValidators.cs]: consider stacking parameters as above.

File: src/Microsoft.IdentityModel.Tokens/InternalValidators.cs [hunk 8]
```diff
@@ -72,7 +72,7 @@ internal static void ValidateLifetimeAndIssuerAfterSignatureNotValidatedJwt(Secu
             }
 
             if (validLifetime && validIssuer)
-                throw LogHelper.LogExceptionMessage(new SecurityTokenSignatureKeyNotFoundException(LogHelper.FormatInvariant(TokenLogMessages.IDX10501, kid, exceptionStrings, securityToken)));
+                throw LogHelper.LogExceptionMessage(new SecurityTokenSignatureKeyNotFoundException(LogHelper.FormatInvariant(TokenLogMessages.IDX10501, kid, LogHelper.MarkAsNonPII(numKeysInTokenValidationParameters), LogHelper.MarkAsNonPII(numKeysInConfiguration), exceptionStrings, securityToken)));
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/InternalValidators.cs]: consider stacking parameters as above.

File: src/Microsoft.IdentityModel.Tokens/InternalValidators.cs [hunk 9]
```diff
@@ -46,7 +46,7 @@ internal static class InternalValidators
         /// <exception cref="SecurityTokenUnableToValidateException">
         /// If the lifetime or issuer are invalid
         /// </exception>
-        internal static void ValidateLifetimeAndIssuerAfterSignatureNotValidatedJwt(SecurityToken securityToken, DateTime? notBefore, DateTime? expires, string kid, TokenValidationParameters validationParameters, BaseConfiguration configuration, StringBuilder exceptionStrings)
+        internal static void ValidateLifetimeAndIssuerAfterSignatureNotValidatedJwt(SecurityToken securityToken, DateTime? notBefore, DateTime? expires, string kid, TokenValidationParameters validationParameters, BaseConfiguration configuration, StringBuilder exceptionStrings, int numKeysInConfiguration, int numKeysInTokenValidationParameters)
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/InternalValidators.cs]: consider stacking parameters as above.

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs (hunk 3 above)
Comment (mafurman) [reply src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs]: I think that makes things a lot more readable. :) Thank you for the suggestion. Fixed! https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1786/commits/27f73791bb1e017be2e7bc1a0b4c25cba31b9a78

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs (hunk 4 above)
Comment (mafurman) [reply src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs]: Fixed! https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1786/commits/27f73791bb1e017be2e7bc1a0b4c25cba31b9a78

File: src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs (hunk 5 above)
Comment (mafurman) [reply src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs]: Fixed! https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1786/commits/27f73791bb1e017be2e7bc1a0b4c25cba31b9a78

File: src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs (hunk 6 above)
Comment (mafurman) [reply src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs]: Fixed! https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1786/commits/27f73791bb1e017be2e7bc1a0b4c25cba31b9a78

File: src/Microsoft.IdentityModel.Tokens/InternalValidators.cs (hunk 7 above)
Comment (mafurman) [reply src/Microsoft.IdentityModel.Tokens/InternalValidators.cs]: Fixed! https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1786/commits/27f73791bb1e017be2e7bc1a0b4c25cba31b9a78

File: src/Microsoft.IdentityModel.Tokens/InternalValidators.cs (hunk 8 above)
Comment (mafurman) [reply src/Microsoft.IdentityModel.Tokens/InternalValidators.cs]: Fixed! https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1786/commits/27f73791bb1e017be2e7bc1a0b4c25cba31b9a78

File: src/Microsoft.IdentityModel.Tokens/InternalValidators.cs (hunk 9 above)
Comment (mafurman) [reply src/Microsoft.IdentityModel.Tokens/InternalValidators.cs]: Fixed! https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1786/commits/27f73791bb1e017be2e7bc1a0b4c25cba31b9a78

**Comment by mafurman** [comment]
@brentschmaltz Let me know if the parameter stacking here looks okay https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1786/commits/27f73791bb1e017be2e7bc1a0b4c25cba31b9a78.

Not sure if we want to have every parameter on a separate line, or if it's okay to group similar parameters together (e.g. strings that are not marked as NonPII).

**Review by brentschmaltz — APPROVED** [review]

**Pull request merged**